include/revng-c/mlir/Dialect/Clift/IR/CliftOps.td

```diff
@@ -141,73 +141,125 @@ def Clift_UndefOp: Clift_Op<"undef"> {
+  let arguments = (ins StrAttr:$sym_name,
+                       TypeAttrOf<Clift_ValueType>:$type);
+
+  let regions = (region Clift_ExpressionRegion:$initializer);
 
   let assemblyFormat = [{
-    $label type($label) attr-dict
+    $type (`=` $initializer^)? attr-dict
   }];
 }
 
-def Clift_MakeLabelOp : Clift_Op<"make_label", [Clift_OneUseOfType<"AssignLabelOp">] > {
-  let results = (outs Res<Clift_LabelType, "", [MemAlloc, MemWrite] >:$result);
+def Clift_MakeLabelOp : Clift_Op<"make_label",
+                                 [Clift_StatementOpInterface,
+                                  Clift_OneUseOfType<"mlir::clift::AssignLabelOp">] > {
+  let results = (outs Clift_LabelType:$result);
 
   let assemblyFormat = [{
     type($result) attr-dict
   }];
+
+  let hasCanonicalizeMethod = 1;
   let hasVerifier = 1;
 
   let builders = [
-     OpBuilder<(ins), [{
-       build($_builder, $_state, mlir::clift::LabelType::get($_builder.getContext()));
-   }]>
-   ];
+    OpBuilder<(ins), [{
+      build($_builder,
+            $_state,
+            mlir::clift::LabelType::get($_builder.getContext()));
+    }]>
+  ];
 }
 
-def Clift_ConstantOp: Clift_Op<"constant", [Pure]> {
-  let arguments = (ins AnyAttrOf<[I64Attr, F64Attr, BoolAttr]>:$value);
+def Clift_ReturnOp : Clift_Op<"return",
+                              [Clift_StatementOpInterface,
+                               Clift_NoFallthrough]> {
+  let regions = (region Clift_ExpressionRegion:$result);
+
+  let assemblyFormat = [{
+    $result attr-dict
+  }];
+}
 
-  let results = (outs Clift_PrimitiveType:$result);
+def Clift_SwitchOp : Clift_Op<"switch",
+                              [Clift_StatementOpInterface]> {
+  let arguments = (ins DenseI64ArrayAttr:$case_values);
+
+  let regions = (region Clift_ExpressionRegion:$condition,
+                        Clift_StatementRegion:$default,
+                        VariadicRegion<Clift_StatementRegion>:$cases);
 
   let assemblyFormat = [{
-    $value type($result) attr-dict
+    $condition $cases (`default` $default^)? attr-dict
   }];
 
-   let builders = [
-        OpBuilder<(ins "mlir::Type":$type, "double":$value), [{
-            build($_builder, $_state, type, $_builder.getF64FloatAttr(value));
-    }]>,OpBuilder<(ins "mlir::Type":$type, "bool":$value), [{
-            build($_builder, $_state, type, $_builder.getBoolAttr(value));
-    }]>,OpBuilder<(ins "mlir::Type":$type, "std::int64_t":$value), [{
-            build($_builder, $_state, type, $_builder.getI64IntegerAttr(value));
-    }]>
-    ];
+  let hasVerifier = 1;
+
+  let builders = [
+    OpBuilder<(ins "llvm::ArrayRef<uint64_t>":$case_values)>
+  ];
+
+  code extraClassDeclaration = [{
+    [[nodiscard]] Region &getConditionRegion() {
+      return getRegion(0);
+    }
+
+    [[nodiscard]] Region &getDefaultCaseRegion() {
+      return getRegion(1);
+    }
+
+    [[nodiscard]] unsigned getNumCases() {
+      return getNumRegions() - 2;
```

Comment:
We should check what e.g. CIR does on structured switches. It's not obvious if you want to count the default as a case or not.